### ISSUE TITLE
mailcap hash change infinite recursion

### ISSUE THREAD

**Issue by wh0** [NixOS/nixpkgs#176286]
### Steps To Reproduce
1. try to build mailcap (needed for python3.9, needed for various other things)

### Build log
```
$ nix-instantiate '<nixpkgs>' -A mailcap
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
/nix/store/2km7xxdp57ylsh8p9jm8zykz2v6cfmbp-mailcap-2.1.53.drv
$ nix-store -r --check /nix/store/2km7xxdp57ylsh8p9jm8zykz2v6cfmbp-mailcap-2.1.53.drv
checking outputs of '/nix/store/2km7xxdp57ylsh8p9jm8zykz2v6cfmbp-mailcap-2.1.53.drv'...

trying https://releases.pagure.org/mailcap/mailcap-2.1.53.tar.xz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 24636  100 24636    0     0  56589      0 --:--:-- --:--:-- --:--:-- 56634
unpacking source archive /build/mailcap-2.1.53.tar.xz
tar: /build/file: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
error: builder for '/nix/store/2km7xxdp57ylsh8p9jm8zykz2v6cfmbp-mailcap-2.1.53.drv' failed with exit code 2
```

### Additional context
https://github.com/NixOS/nixpkgs/pull/173430 changes what `postFetch` means in `fetchzip`; mailcap https://github.com/NixOS/nixpkgs/blob/master/pkgs/data/misc/mailcap/default.nix#L13 expects the downloaded file still to be in `$downloadedFile`, but the new `fetchzip` has moved it to `$renamed` https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchzip/default.nix#L47.

looks like mailcap's postFetch really does intend to replace fetchzip's. maybe it would be better to change to plain fetchurl?

### Notify maintainers
@c0bw3b @Artturin , if I may

### Metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.109+, Debian GNU/Linux, 11 (bullseye), nobuild`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.9.0`
 - channels(...): `"nixpkgs"`
 - nixpkgs: `/home/.../.nix-defexpr/channels/nixpkgs`
```

(some info redacted)

**Comment by Artturin** [comment]
this has exposed a larger problem which is infinite recursions

```nix
error: infinite recursion encountered

       at //builtin/derivation.nix:19:19:
       … while evaluating 'optionalString'

       at /nix/store/1ljgyxz46vh768ya2djkmf3cqsn3ln1p-source/lib/strings.nix:203:5:

          202|     # String to return if condition is true
          203|     string: if cond then string else "";
             |     ^
          204|

       … from call site

       at /nix/store/1ljgyxz46vh768ya2djkmf3cqsn3ln1p-source/pkgs/development/interpreters/python/cpython/default.nix:278:8:

          277|       --replace "'/bin/sh'" "'${bash}/bin/sh'"
          278|   '' + optionalString mimetypesSupport ''
             |        ^
          279|     substituteInPlace Lib/mimetypes.py \

       … while evaluating the attribute 'postPatch' of the derivation 'python3-3.9.13'

       at /nix/store/1ljgyxz46vh768ya2djkmf3cqsn3ln1p-source/pkgs/stdenv/generic/make-derivation.nix:278:7:

          277|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          278|       name =
             |       ^
          279|         let

       … while evaluating the attribute 'nativeBuildInputs' of the derivation 'python3.9-buildcatrust-0.1.3'

       at /nix/store/1ljgyxz46vh768ya2djkmf3cqsn3ln1p-source/pkgs/stdenv/generic/make-derivation.nix:278:7:

          277|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          278|       name =
             |       ^
          279|         let

       … while evaluating the attribute 'nativeBuildInputs' of the derivation 'nss-cacert-3.77'

       at /nix/store/1ljgyxz46vh768ya2djkmf3cqsn3ln1p-source/pkgs/stdenv/generic/make-derivation.nix:278:7:

          277|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          278|       name =
             |       ^
          279|         let

       … while evaluating the attribute 'SSL_CERT_FILE' of the derivation 'mailcap-2.1.53'

       at /nix/store/1ljgyxz46vh768ya2djkmf3cqsn3ln1p-source/pkgs/stdenv/generic/make-derivation.nix:278:7:

          277|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          278|       name =
             |       ^
          279|         let
```

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/176291

**Comment by Artturin** [comment]
where are you getting this error so we'll know if this needs to go in to master

**Comment by wh0** [comment]
nixpkgs unstable and nixos 22.05

**Comment by SuperSandro2000** [comment]
Fix got reverted in #177541

**Comment by dotlambda** [comment]
Wasn't this solved by switching to fetchurl?
